File: player-mfe/src/entities/team.ts
import { TeamResponse } from "../types/requests";

class Team {
  private name: string;
  private score: number;
  private isPositiveHalf: boolean;
  private color: RobotColor;
  private goalkeeperId: number;
  private yellowCards: number;
  private timeToExpireActiveYellowCards: number[];
  private redCards: number;
  private timeoutsLeft: number;
  private totalTimeoutTimeLeft: Date;
  private foulsCommitted: number;
  private consecutiveBallPlacementFailures: number;
  private isBallPlacementEnabled: boolean;
  private hasBallPlacementFailuresReachedMaximum: boolean;
  private maximumAllowedRobots: number;
  private isRobotSubstitutionRequested: boolean;
  private isRobotSubstitutionAllowed: boolean;
  private robotSubstitutionsLeft: number;
  private robotSubstitutionTimeLeft: Date;

  constructor(data: TeamResponse) {
    this.name = data.name;
    this.score = data.score;
    this.isPositiveHalf = data.is_positive_half;
    this.color = data.robot_color;
    this.goalkeeperId = data.goalkeeper_id;
    this.yellowCards = data.yellow_cards;
    this.timeToExpireActiveYellowCards =
      data.time_to_expire_active_yellow_cards;
    this.redCards = data.red_cards;
    this.timeoutsLeft = data.timeouts_left;
    this.totalTimeoutTimeLeft = new Date(data.total_timeout_time_left);
    this.foulsCommitted = data.fouls_committed;
    this.consecutiveBallPlacementFailures =
      data.consecutive_ball_placement_failures;
    this.isBallPlacementEnabled = data.is_ball_placement_enabled;
    this.hasBallPlacementFailuresReachedMaximum =
      data.has_ball_placement_failures_reached_maximum;
    this.maximumAllowedRobots = data.maximum_allowed_robots;
    this.isRobotSubstitutionRequested = data.is_robot_substitution_requested;
    this.isRobotSubstitutionAllowed = data.is_robot_substitution_allowed;
    this.robotSubstitutionsLeft = data.robot_substitutions_left;
    this.robotSubstitutionTimeLeft = new Date(
      data.robot_substitution_time_left
    );
  }
}

export default Team;

File: player-bff/player-sonson/internal/entity/team.go
package entity

import (
	"time"

	"github.com/robocin/ssl-core/protocols/playback"
)

type Team struct {
	Name                                   string          `json:"name"`
	Score                                  uint32          `json:"score"`
	RobotColor                             string          `json:"robot_color"`
	GoalkeeperId                           uint32          `json:"goalkeeper_id"`
	IsPositiveHalf                         bool            `json:"is_positive_half"`
	YellowCards                            uint32          `json:"yellow_cards"`
	TimeToExpireActiveYellowCards          []time.Duration `json:"time_to_expire_active_yellow_cards"`
	RedCards                               uint32          `json:"red_cards"`
	TimeoutsLeft                           uint32          `json:"timeouts_left"`
	TotalTimeoutTimeLeft                   time.Duration   `json:"total_timeout_time_left"`
	FoulsCommited                          uint32          `json:"fouls_committed"`
	ConsecutiveBallPlacementFailures       uint32          `json:"consecutive_ball_placement_failures"`
	IsBallPlacementEnabled                 bool            `json:"is_ball_placement_enabled"`
	HasBallPlacementFailuresReachedMaximum bool            `json:"has_ball_placement_failures_reached_maximum"`
	MaximumAllowedRobots                   uint32          `json:"maximum_allowed_robots"`
	IsRobotSubstitutionRequested           bool            `json:"is_robot_substitution_requested"`
	IsRobotSubstitutionAllowed             bool            `json:"is_robot_substitution_allowed"`
	RobotSubstitutionsLeft                 uint32          `json:"robot_substitutions_left"`
	RobotSubstitutionTimeLeft              time.Duration   `json:"robot_substitution_time_left"`
}

func NewTeam(team *playback.GameStatus_Team, positiveHalf bool) Team {
	color := "yellow"
	if len(team.GetRobotIds()) > 0 {
		color = translateColor(team.GetRobotIds()[0].GetColor())
	}

	times := make([]time.Duration, 0)
	for _, card := range team.GetTimeToExpireActiveYellowCards() {
		times = append(times, card.AsDuration())
	}

	return Team{
		Name:                                   team.GetName(),
		Score:                                  team.GetScore(),
		IsPositiveHalf:                         positiveHalf,
		RobotColor:                             color,
		GoalkeeperId:                           team.GetGoalkeeperId().GetNumber(),
		YellowCards:                            team.GetYellowCards(),
		TimeToExpireActiveYellowCards:          times,
		RedCards:                               team.GetRedCards(),
		TimeoutsLeft:                           team.GetTimeoutsLeft(),
		TotalTimeoutTimeLeft:                   team.GetTotalTimeoutTimeLeft().AsDuration(),
		FoulsCommited:                          team.GetTotalFoulsCommitted(),
		ConsecutiveBallPlacementFailures:       team.GetConsecutiveBallPlacementFailures(),
		IsBallPlacementEnabled:                 team.GetIsBallPlacementEnabled(),
		HasBallPlacementFailuresReachedMaximum: team.GetHasBallPlacementFailuresReachedMaximum(),
		MaximumAllowedRobots:                   team.GetMaximumAllowedRobots(),
		IsRobotSubstitutionRequested:           team.GetIsRobotSubstitutionRequested(),
		IsRobotSubstitutionAllowed:             team.GetIsRobotSubstitutionAllowed(),
		RobotSubstitutionsLeft:                 team.GetRobotSubstitutionsLeft(),
		RobotSubstitutionTimeLeft:              team.GetRobotSubstitutionTimeLeft().AsDuration(),
	}
}
